### PR TITLE
MapperService has to be passed in as null for EnginePlugins CodecService constructor

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
Introduces `CodecServiceFactory` and extends the `EnginePlugin` with new method (the existing `getCustomCodecService` is marked as `@Deprecated`).

```
    default Optional<CodecServiceFactory> getCustomCodecServiceFactory() {
        return Optional.empty();
    }
```

The factory accepts `CodecServiceConfig` instance which provides more context to the `CodecService` creators, including `MapperService` is available for the particular index.

### Issues Resolved
Closes https://github.com/opensearch-project/OpenSearch/issues/1805
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:80]: This seems to be a bug: the `IllegalStateException` is going to be throw even if `enginePlugin` does not provide any codec service (returns `Optional.empty`).

File: server/src/main/java/org/opensearch/plugins/EnginePlugin.java
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: @dblock @andrross @nknize what do you think about this approach guys? this is non-breaking change but opens two ways to configure `CodecService` now (for 2.0.0 we could drop `getCustomCodecService` actually)

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (andrross) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:80]: Seems bad. Has this case just never been hit because in practice there is only ever 1 engine plugin or because the last plugin in the list is the one to override the codec service?

Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:80]: Hard to say, but I would guess that only handful of engine plugins do provided codec service (and probably there is usually only 1 as you mentioned).

File: server/src/main/java/org/opensearch/plugins/EnginePlugin.java
Comment (reta) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: Thanks @andrross ! Yes, super trivial: 

```
Optional<CodecService> getCustomCodecService(IndexSettings indexSettings) {
    return getCustomCodecServiceFactory().map(f -> f.createCodecService(new CodecServiceConfig(indexSettings, null));
}
```

The `CodecServiceConfig` also gives an opportunity to include more things in the future (if necessary).

**Comment by andrross** [comment]
@jmazanec15 Can you take a look at this and confirm it will solve the problem you described in #1805?

**Comment by jmazanec15** [comment]
@andrross Apologies, just saw this tag. Will take a look this week.

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (jmazanec15) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:80]: Right, and I believe this change was introduced in 1.2 as well

File: server/src/main/java/org/opensearch/indices/IndicesService.java
Comment (jmazanec15) [inline server/src/main/java/org/opensearch/indices/IndicesService.java:745]: When would IndexService be null? If it is should it fail?

File: server/src/main/java/org/opensearch/plugins/EnginePlugin.java
Comment (jmazanec15) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: Why doesnt this take indexSettings as an argument? If I have an engine plugins implementing this method and want to create an unrelated index, wouldn't this getCustomCodecServiceFactory still be used?

Comment (jmazanec15) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: Could this be an issue: https://github.com/opensearch-project/OpenSearch/pull/2177/files#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R113? If index setting parsing is done after this check?

File: server/src/main/java/org/opensearch/indices/IndicesService.java
Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:745]: This is good question - I cannot say exactly, but since `indices` is just map, the `get` may return `null`, better be ready for it.

File: server/src/main/java/org/opensearch/plugins/EnginePlugin.java
Comment (reta) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: @jmazanec15 it does not take the `indexSettings` because it just a factory, but the factory's method `createCodecService`, exactly where the codec service instance is expected to be created, takes the `indexSettings` as part of `CodecServiceConfig`. It also opens the door to add more services in the future without changing the method signatures.

Comment (jmazanec15) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: But wouldnt this scenario fail: I have EnginePlugin1 and EnginePlugin2 that both return non-null for getCustomCodecServiceFactory. There is an index setting, "engine_plugin_1" :"true/false" and "engine_plugin_2" :"true/false".

I create an index with  "engine_plugin_1" :"true" and "engine_plugin_2" :"false". Wouldnt this fail here: https://github.com/opensearch-project/OpenSearch/pull/2177/files#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R103?

Comment (reta) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java:83]: Ah I see what you mean, yeah, I think we could add `indexSettings` here so the plugin could decide it the factory should be provided or not.

Comment (reta) [reply server/src/main/java/org/opensearch/plugins/EnginePlugin.java]: @jmazanec15 done, thanks for the feedback!

**Review by andrross — APPROVED** [review]
LGTM! I'll just wait for sign off from @jmazanec15 before merging.

File: server/src/main/java/org/opensearch/indices/IndicesService.java
Comment (jmazanec15) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:745]: I think my opinion is that we should only create a codec service in one place and it should be here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L324. We should pass in the CodecService factory there and get rid of the codec service in EngineConfigFactory.

File: server/src/main/java/org/opensearch/index/codec/CodecServiceConfig.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/codec/CodecServiceConfig.java:21]: We could, I believe the presence of the index settings could be beneficial, otherwise would leave that implementors to pass the index settings from `EnginePlugin` to `CodecService` instance if needed

File: server/src/main/java/org/opensearch/indices/IndicesService.java
Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:745]: Very good point, looking into it now, thank you!

Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: @jmazanec15 delayed the CodeService creation up to the moment shard is initialized, `MapperService` is taken directly from there (if available). Sadly, I had to add new method `newEngineConfig` and deprecate the old one in order to keep backward compatibility.

Comment (jmazanec15) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: Could we do something like this: https://github.com/jmazanec15/OpenSearch/commit/165f50a3bf9fa067f777acd85770608299362534? This way, we can exclusively use CodecServiceFactory to build codec and not change any function interfaces?

Alternatively, I thought about building CodecService directly in newEngineConfig. In this case, we would switch from passing in CodecService directly to passing in CodecServiceConfig. But, I thought this might cause us to create more CodecServices than we need to.

Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: @jmazanec15 hm ... to me, the logger service does not belong here, may be if you share use case, we could think about it

Comment (jmazanec15) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: @reta Logger is one of the constructor arguments to CodecService. I think the CodecServiceConfig will need to have it to build CodecService - similar to MapperService. That way, we can use CodecServiceFactory to build default CodecService.

Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: @jmazanec15 I should have looked it :) Thank you, will do the change shortly

Comment (reta) [reply server/src/main/java/org/opensearch/indices/IndicesService.java]: Done, thank @jmazanec15 !

**Comment by dblock** [comment]
@jmazanec15 you're good with this?

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (jmazanec15) [inline server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:216]: I think we should only build CodecService in one place and always use the factory. That way, we can avoid ternary statements like the above. For default CodecServiceFactory, we can set it like I did [here](https://github.com/jmazanec15/OpenSearch/commit/165f50a3bf9fa067f777acd85770608299362534#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R119).

Im somewhat conflicted on where to build CodecService. I think there are 2 options: 

#### 1. Call `this.codecServiceFactory.createCodecService` here and dont take in CodecService as arg to newEngineConfig

The pro's to this approach are that we can hide details about CodecService from all other parts of code - in IndexShard we dont need to do anything with codec service. Its all handled by EngineConfigFactory. 

The cons are that a new CodecService would get created every time this method gets called. I'm not really sure if there will be side effects to this. In IndexShard, this gets called in `resetEngineToGlobalCheckpoint` and `innerOpenEngineAndTranslog`. One potential issue could arise if a user changes IndexSettings between these 2 operations and that impacts how the CodecServiceFactory builds the CodecService (maybe this is a feature not a bug?).

#### 2. Call it in IndexShard's constructor by exposing getter on EngineConfigFactory's codecServiceFactory

Alternatively, we could create the codec service in IndexShards constructor. This is how it is being done now for the default. To do this, we would need to expose a getter method for EngineConfigFactory to get its codecServiceFactory and then call `this.codecServiceFactory.createCodecService()`. Then, we wouldnt change anything in `EngineConfigFactory.newEngineConfig()`. The getter, however, feels a little bit out of place here. I took this approach [here](https://github.com/jmazanec15/OpenSearch/commit/165f50a3bf9fa067f777acd85770608299362534#diff-acf4481a07fc77c1b5d6eaca20a6a1ce1b171615c4349b67bd23b2142a2f0ccbR330). 

I think I am leaning towards approach 2 as opposed to 1, to keep consistent with how its being done now.

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java:92]: This is a public method which could be used (by plugins fe) so we should not break them

File: server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java]: @jmazanec15 I think I got your idea, the challenge here is to preserve the semantics of the `CodecService` instantiation, I just pushed the solution which I believe is the best effort: 
 - new method `EngineConfigFactory::newCodecServiceOrDefault` which basically either create a new `CodecService` instance from factory or uses the existing one
 - `IndexShard::newEngineConfig` uses  `EngineConfigFactory::newCodecServiceOrDefault` to pass the `CodecService` instance to `EngineConfigFactory::newEngineConfig`
 - `EngineConfigFactory::newEngineConfig` has the fallback in case passed `CodecService` is `null` but there is `CodecServiceFactory` configured - the old way to create `CodecService` without `MapperService` and `Logger`

The APIs stay unchanged, thank you

Comment (jmazanec15) [reply server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java]: That makes sense thanks!

**Comment by dblock** [comment]
If @jmazanec15 is good with this change we can merge.

**Comment by jmazanec15** [comment]
Looks good @dblock

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**
